### PR TITLE
Feature/pipeline generator GitHub

### PR DESCRIPTION
**Pull request by ultymatom**
This PR is for the creation of the pipeline generator for github actions.

This also include changes in the pipeline generator for azure, as some funcitons are common I created a lib file

### CODE REVIEW

File: scripts/pipelines/github/pipeline_generator.sh [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+eval set -- "$FLAGS"
+while true; do
+    case "$1" in
+        -c | --config-file)       configFile=$2; shift 2;;
+        -n | --pipeline-name)     export pipelineName=$2; shift 2;;
+        -d | --local-directory)   localDirectory=$2; shift 2;;
+        -a | --artifact-path)     artifactPath=$2; shift 2;;
+        -b | --target-branch)     targetBranch=$2; shift 2;;
+        -l | --language)          language=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:13]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        --build-pipeline-name)    export buildPipelineName=$2; shift 2;;
+        --sonar-url)              sonarUrl=$2; shift 2;;
+        --sonar-token)            sonarToken=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:16]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        -i | --image-name)        imageName=$2; shift 2;;
+        -u | --registry-user)     dockerUser=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:18]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        -p | --registry-password) dockerPassword=$2; shift 2;;
+        --resource-group)         resourceGroupName=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:20]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        --storage-account)        storageAccountName=$2; shift 2;;
+        --storage-container)      storageContainerName=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:22]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        --cluster-name)           clusterName=$2; shift 2;;
+        --s3-bucket)              s3Bucket=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:24]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+        --s3-key-path)            s3KeyPath=$2; shift 2;;
+        --quality-pipeline-name)  export qualityPipelineName=$2; shift 2;;
+        --test-pipeline-name)     export testPipelineName=$2; shift 2;;
+        --dockerfile)             dockerFile=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:28]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 8]
```diff
@@ -0,0 +1,228 @@
+#!/bin/bash
+set -e
+FLAGS=$(getopt -a --options c:n:d:a:b:l:i:u:p:hw --long "config-file:,pipeline-name:,local-directory:,artifact-path:,target-branch:,language:,build-pipeline-name:,sonar-url:,sonar-token:,image-name:,registry-user:,registry-password:,resource-group:,storage-account:,storage-container:,cluster-name:,s3-bucket:,s3-key-path:,quality-pipeline-name:,dockerfile:,test-pipeline-name:,aws-access-key:,aws-secret-access-key:,aws-region:,help" -- "$@")
+
+eval set -- "$FLAGS"
+while true; do
+    case "$1" in
+        -c | --config-file)       configFile=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:8]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+        -n | --pipeline-name)     export pipelineName=$2; shift 2;;
+        -d | --local-directory)   localDirectory=$2; shift 2;;
+        -a | --artifact-path)     artifactPath=$2; shift 2;;
+        -b | --target-branch)     targetBranch=$2; shift 2;;
+        -l | --language)          language=$2; shift 2;;
+        --build-pipeline-name)    export buildPipelineName=$2; shift 2;;
+        --sonar-url)              sonarUrl=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:15]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        --sonar-token)            sonarToken=$2; shift 2;;
+        -i | --image-name)        imageName=$2; shift 2;;
+        -u | --registry-user)     dockerUser=$2; shift 2;;
+        -p | --registry-password) dockerPassword=$2; shift 2;;
+        --resource-group)         resourceGroupName=$2; shift 2;;
+        --storage-account)        storageAccountName=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:21]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+        --storage-container)      storageContainerName=$2; shift 2;;
+        --cluster-name)           clusterName=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:23]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+        --s3-bucket)              s3Bucket=$2; shift 2;;
+        --s3-key-path)            s3KeyPath=$2; shift 2;;
+        --quality-pipeline-name)  export qualityPipelineName=$2; shift 2;;
+        --test-pipeline-name)     export testPipelineName=$2; shift 2;;
+        --dockerfile)             dockerFile=$2; shift 2;;
+        --aws-access-key)         awsAccessKey="$2"; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:29]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+        --aws-secret-access-key)  awsSecretAccessKey="$2"; shift 2;;
+        --aws-region)             awsRegion="$2"; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:31]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+hangarPath=${pipelineGeneratorFullPath/$pipelineGeneratorRepoPath}
+}
+
+# Function that adds the variables to be used in the pipeline.
+function addCommonPipelineVariables {
+    if test -z ${artifactPath}
+    then
+        echo "Skipping creation of the variable artifactPath as the flag has not been used."
+        # Delete the commentary to set the artifactPath input/var
+        sed -i '/# mark to insert additional artifact input #/d' "${localDirectory}/${pipelinePath}/${yamlFile}"
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:137]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+        --storage-account)        storageAccountName=$2; shift 2;;
+        --storage-container)      storageContainerName=$2; shift 2;;
+        --cluster-name)           clusterName=$2; shift 2;;
+        --s3-bucket)              s3Bucket=$2; shift 2;;
+        --s3-key-path)            s3KeyPath=$2; shift 2;;
+        --quality-pipeline-name)  export qualityPipelineName=$2; shift 2;;
+        --test-pipeline-name)     export testPipelineName=$2; shift 2;;
+        --dockerfile)             dockerFile=$2; shift 2;;
+        --aws-access-key)         awsAccessKey="$2"; shift 2;;
+        --aws-secret-access-key)  awsSecretAccessKey="$2"; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:30]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+done
+
+# Colours for the messages.
+white='\e[1;37m'
+green='\e[1;32m'
+red='\e[0;31m'
+
+# Loading common function
+# Common var
+commonTemplatesPath="scripts/pipelines/github/templates/common"
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:45]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh (hunk 14 above)
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:137]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+function addCommonPipelineVariables {
+    if test -z ${artifactPath}
+    then
+        echo "Skipping creation of the variable artifactPath as the flag has not been used."
+        # Delete the commentary to set the artifactPath input/var
+        sed -i '/# mark to insert additional artifact input #/d' "${localDirectory}/${pipelinePath}/${yamlFile}"
+        sed -i '/# mark to insert additional artifact env var #/d' "${localDirectory}/${pipelinePath}/${yamlFile}"
+    else
+        # add the input for the additional artifact
+        grep "    inputs:" ${localDirectory}/${pipelinePath}/${yamlFile} > /dev/null && textArtifactPathInput="      artifactPath:\n       required: false\n       default: ${artifactPath//\//\\/}"
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:141]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+        sed -i "s/# mark to insert additional artifact env var #/$textArtifactPathVar/" "${localDirectory}/${pipelinePath}/${yamlFile}"
+    fi
+}
+
+function createPR {
+    # Check if a target branch is supplied.
+    if test -z "$targetBranch"
+    then
+        # No branch specified in the parameters, no Pull Request is created, the code will be stored in the current branch.
+        echo -e "${green}No branch specified to do the Pull Request, changes left in the ${sourceBranch} branch."
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:157]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+        fi
+    fi
+}
+
+if [[ "$help" == "true" ]]; then help; fi
+
+obtainHangarPath
+
+# Load common functions
+. "$hangarPath/scripts/pipelines/common/pipeline_generator.lib"
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:201]: @sonatype-lift ignore

File: scripts/pipelines/azure-devops/pipeline_generator.sh [hunk 20]
```diff
@@ -274,12 +197,17 @@ function createPR {
 
 if [[ "$help" == "true" ]]; then help; fi
 
+obtainHangarPath
+
+# Load common functions
+. "$hangarPath/scripts/pipelines/common/pipeline_generator.lib"
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/azure-devops/pipeline_generator.sh:203]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+        -d | --local-directory)   localDirectory=$2; shift 2;;
+        -a | --artifact-path)     artifactPath=$2; shift 2;;
+        -b | --target-branch)     targetBranch=$2; shift 2;;
+        -l | --language)          language=$2; shift 2;;
+        --build-pipeline-name)    export buildPipelineName=$2; shift 2;;
+        --sonar-url)              sonarUrl=$2; shift 2;;
+        --sonar-token)            sonarToken=$2; shift 2;;
+        -i | --image-name)        imageName=$2; shift 2;;
+        -u | --registry-user)     dockerUser=$2; shift 2;;
+        -p | --registry-password) dockerPassword=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:19]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+        -n | --pipeline-name)     export pipelineName=$2; shift 2;;
+        -d | --local-directory)   localDirectory=$2; shift 2;;
+        -a | --artifact-path)     artifactPath=$2; shift 2;;
+        -b | --target-branch)     targetBranch=$2; shift 2;;
+        -l | --language)          language=$2; shift 2;;
+        --build-pipeline-name)    export buildPipelineName=$2; shift 2;;
+        --sonar-url)              sonarUrl=$2; shift 2;;
+        --sonar-token)            sonarToken=$2; shift 2;;
+        -i | --image-name)        imageName=$2; shift 2;;
+        -u | --registry-user)     dockerUser=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:18]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+        -p | --registry-password) dockerPassword=$2; shift 2;;
+        --resource-group)         resourceGroupName=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:20]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 24]
```diff
@@ -0,0 +1,229 @@
...
+        --storage-account)        storageAccountName=$2; shift 2;;
+        --storage-container)      storageContainerName=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:22]: @sonatype-lift ignore

File: scripts/pipelines/github/pipeline_generator.sh [hunk 25]
```diff
@@ -0,0 +1,229 @@
...
+        -b | --target-branch)     targetBranch=$2; shift 2;;
+        -l | --language)          language=$2; shift 2;;
+        --build-pipeline-name)    export buildPipelineName=$2; shift 2;;
+        --sonar-url)              sonarUrl=$2; shift 2;;
+        --sonar-token)            sonarToken=$2; shift 2;;
+        -i | --image-name)        imageName=$2; shift 2;;
+        -u | --registry-user)     dockerUser=$2; shift 2;;
+        -p | --registry-password) dockerPassword=$2; shift 2;;
+        --resource-group)         resourceGroupName=$2; shift 2;;
+        --storage-account)        storageAccountName=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:21]: @sonatype-lift ignore

File: scripts/pipelines/gitlab/pipeline_generator.sh [hunk 26]
```diff
@@ -0,0 +1,208 @@
...
+    case "$1" in
+        -c | --config-file)       configFile=$2; shift 2;;
+        -n | --pipeline-name)     export pipelineName=$2; shift 2;;
+        -d | --local-directory)   localDirectory=$2; shift 2;;
+        -a | --artifact-path)     artifactPath=$2; shift 2;;
+        -b | --target-branch)     targetBranch=$2; shift 2;;
+        -l | --language)          language=$2; shift 2;;
+        --build-pipeline-name)    export buildPipelineName=$2; shift 2;;
+        --sonar-url)              sonarUrl=$2; shift 2;;
+        --sonar-token)            sonarToken=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/gitlab/pipeline_generator.sh:16]: @sonatype-lift ignore

File: scripts/pipelines/gitlab/pipeline_generator.sh [hunk 27]
```diff
@@ -0,0 +1,208 @@
...
+        -i | --image-name)        imageName=$2; shift 2;;
+        -u | --registry-user)     dockerUser=$2; shift 2;;
+        -p | --registry-password) dockerPassword=$2; shift 2;;
+        --resource-group)         resourceGroupName=$2; shift 2;;
```
[the remark this answers is not in the document]
Comment (ultymatom) [reply scripts/pipelines/gitlab/pipeline_generator.sh:20]: @sonatype-lift ignore

File: scripts/pipelines/gitlab/templates/build/quarkus-build.sh [hunk 28]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
```
Comment (albertdb) [inline scripts/pipelines/gitlab/templates/build/quarkus-build.sh:1]: This does not belong to this PR.

File: scripts/pipelines/gitlab/templates/build/build-pipeline.cfg [hunk 29]
```diff
@@ -0,0 +1,16 @@
+# Mandatory flags.
```
Comment (albertdb) [inline scripts/pipelines/gitlab/templates/build/build-pipeline.cfg:1]: This does not belong here nor to this PR.

File: scripts/pipelines/gitlab/pipeline_generator.sh [hunk 30]
```diff
@@ -0,0 +1,208 @@
+#!/bin/bash
```
Comment (albertdb) [inline scripts/pipelines/gitlab/pipeline_generator.sh:1]: This does not belong to this PR and it is a copy-paste from GitHub one.

File: scripts/pipelines/github/pipeline_generator.sh [hunk 31]
```diff
@@ -0,0 +1,207 @@
...
+green='\e[1;32m'
+red='\e[0;31m'
+
+# Common var
+commonTemplatesPath="scripts/pipelines/github/templates/common" # Path for common files of the pipelines
+pipelinePath=".github/workflows" # Path to the pipelines.
+scriptFilePath=".github/workflows/scripts" # Path to the scripts.
+export provider="github"
+
+function help {
```
Comment (albertdb) [inline scripts/pipelines/github/pipeline_generator.sh:49]: Can't we move help and flag parsing to common? We can create provider specific section and parsing if needed, right?

**Review by albertdb — CHANGES_REQUESTED** [review]

File: scripts/pipelines/github/pipeline_generator.sh (hunk 31 above)
Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh:49]: The problem is that the help function is supposed to be executed before anything if the user ask for it

Comment (ultymatom) [reply scripts/pipelines/github/pipeline_generator.sh]: my bad we can do it, and it's done

**Review by albertdb — APPROVED** [review]

**Review by albertdb — APPROVED** [review]

**Pull request merged**